pkg/dockerregistry/server/writelimiter/blobstore.go

```diff
@@ -0,0 +1,119 @@
+package writelimiter
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+
+	"github.com/openshift/origin/pkg/dockerregistry/server/types"
+)
+
+// ErrResourcesExhausted is returned when no more BlobWriters can be created
+// because of the limit.
+var ErrResourcesExhausted = errors.New("writers limit is reached")
+
+type writerBouncer struct {
+	sem  *semaphore
+	once sync.Once
+}
+
+func newWriterBouncer(sem *semaphore) (*writerBouncer, error) {
+	if !sem.TryDown() {
+		return nil, ErrResourcesExhausted
+	}
+	return &writerBouncer{
+		sem: sem,
+	}, nil
+}
+
+func (g *writerBouncer) Release() {
+	g.once.Do(g.sem.Up)
+}
+
+type blobWriter struct {
+	distribution.BlobWriter
+	bouncer *writerBouncer
+}
+
+var _ distribution.BlobWriter = blobWriter{}
+
+func (bw blobWriter) Close() error {
+	bw.bouncer.Release()
+	return bw.BlobWriter.Close()
+}
+
+func (bw blobWriter) Commit(ctx context.Context, desc distribution.Descriptor) (distribution.Descriptor, error) {
+	bw.bouncer.Release()
+	return bw.BlobWriter.Commit(ctx, desc)
+}
+
+func (bw blobWriter) Cancel(ctx context.Context) error {
+	bw.bouncer.Release()
+	return bw.BlobWriter.Cancel(ctx)
+}
+
+type blobStore struct {
+	distribution.BlobStore
+	sem *semaphore
+}
+
+var _ distribution.BlobStore = blobStore{}
+
+func (bs blobStore) Create(ctx context.Context, options ...distribution.BlobCreateOption) (distribution.BlobWriter, error) {
+	bouncer, err := newWriterBouncer(bs.sem)
```

Comment:
> Why not make this a wrapper BlobStore?

BlobStore doesn't have a method that would be called on destruction.

> Why isn't this is a wrapper around individual block allocations?

Block allocations happen inside the upstream S3 storage driver. But suppose I will wrap all `append` calls there to catch allocations, then how can I catch deallocations?

> When we have a blob that is only 20-50k to upload (manifests, small layers), do we still allocate a full chunk?

No, we don't.

Distribution stores manifests using `PutContent(..., content []byte)` method. And as it already gets all content in the buffer, there is no need in other temporary buffers. But distribution reads the whole manifest into memory before it calls PutContent.

For layers it uses roughly this writer:

```
struct writer {
    buf []byte
    chunkSize int // 10 MB by default
}

func (w *writer) Write(b []byte) (int, error) {
    w.buf = append(w.buf, b...)
    if len(w.buf) > w.chunkSize {
        sendPartToS3(w.buf) // only the last part can be less than 5 MB
        w.buf = nil
    }
    return len(b), nil
}
```

So the whole layer will be read into memory if the layer is small enough. The exact amount of memory depends on `io.Copy` and `append` implementations.

In the real writer there is also a second buffer, because it tries to make the last part no less than 5 MB (if the last chunk is more that 5 MB, another part of data can be added to the blob). Virtually it means it tries to keep 10 MB (by default) of data which should be flushed on Close.

> If minimum chunk size is 5mb, that means 200 uploads per gig.

In practice this number will be even lower because of the second buffer.